cmd/zpool/zpool_vdev.c

```diff
@@ -423,7 +423,7 @@ check_disk(const char *path, blkid_cache cache, int force,
 	if (!iswholedisk)
 		return (check_slice(path, cache, force, isspare));
 
-	if ((fd = open(path, O_RDONLY|O_DIRECT|O_EXCL)) < 0) {
+	if ((fd = open(path, O_RDONLY|O_DIRECT)) < 0) {
```

Comment:
It seems `O_EXCL` is only buying us a slightly better error message here, with and without `-f`:

System status (pools and mounted fs on `/dev/sdc1`):
```
root@linux:/usr/src/zfs# zpool status 
  pool: p1
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	p1          ONLINE       0     0     0
	  sda       ONLINE       0     0     0

errors: No known data errors

  pool: p2
 state: ONLINE
  scan: none requested
config:

	NAME        STATE     READ WRITE CKSUM
	p2          ONLINE       0     0     0
	  sdb       ONLINE       0     0     0

errors: No known data errors
root@linux:/usr/src/zfs# df -h /dev/sdc1
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdc1       120M  1.6M  110M   2% /mnt
root@linux:/usr/src/zfs# 
```

Previous behaviour:
```
root@linux:/usr/src/zfs# zpool add p1 spare sdc
/dev/sdc is in use and contains a unknown filesystem.
root@linux:/usr/src/zfs# zpool add p1 spare sdc1
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 contains a filesystem of type 'ext4'
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# zpool add -f p1 spare sdc
/dev/sdc is in use and contains a unknown filesystem.
root@linux:/usr/src/zfs# zpool add -f p1 spare sdc1
cannot open '/dev/sdc1': Device or resource busy
cannot add to 'p1': one or more vdevs refer to the same device
root@linux:/usr/src/zfs# 
```

Patched behaviour:
```
root@linux:/usr/src/zfs# zpool add p1 spare sdc
cannot label 'sdc': cannot label '/dev/sdc': unable to open device: 16
root@linux:/usr/src/zfs# zpool add p1 spare sdc1
invalid vdev specification
use '-f' to override the following errors:
/dev/sdc1 contains a filesystem of type 'ext4'
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# 
root@linux:/usr/src/zfs# zpool add -f p1 spare sdc
cannot label 'sdc': cannot label '/dev/sdc': unable to open device: 16
root@linux:/usr/src/zfs# zpool add -f p1 spare sdc1
cannot open '/dev/sdc1': Device or resource busy
cannot add to 'p1': one or more vdevs refer to the same device
root@linux:/usr/src/zfs#
```

I am going to try and keep the error "_is in use and contains a unknown filesystem_", though if i had to choose i would prefer a misleading error message (and the optional "documentation" issue requesting to improve it) rather than this missing functionality (shared spare).

Also it seems this could use a new test in the "inuse" group.

Reply:
Agree, it'd be nice to have both but we can live with a slightly worse error message to restore the missing functionality.